contrib/hadoop/src/test/java/com/google/cloud/dataflow/contrib/hadoop/HadoopFileSourceTest.java

```diff
@@ -0,0 +1,123 @@
+
+import com.google.cloud.dataflow.sdk.io.BoundedSource;
+import com.google.cloud.dataflow.sdk.io.SourceTestUtils;
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.options.PipelineOptionsFactory;
+import com.google.cloud.dataflow.sdk.values.KV;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.IOUtils;
+import org.apache.hadoop.io.IntWritable;
+import org.apache.hadoop.io.SequenceFile;
+import org.apache.hadoop.io.SequenceFile.Writer;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.mapreduce.lib.input.SequenceFileInputFormat;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class HadoopFileSourceTest {
+
+  private static final int DEFAULT_RECORD_COUNT = 10000;
+
+  @Rule
+  public TemporaryFolder tmpFolder = new TemporaryFolder();
+
+  @Test
+  public void testRead() throws Exception {
+    List<KV<IntWritable, Text>> expected = createRandomRecords();
+    String filename = generateTestFile("tmp.seq", expected);
+
+    HadoopFileSource<IntWritable, Text> source =
+        HadoopFileSource.from(filename, SequenceFileInputFormat.class,
+            IntWritable.class, Text.class);
+
+    assertEquals(new File(filename).length(), source.getEstimatedSizeBytes(null));
+
+    List<KV<IntWritable, Text>> actual = SourceTestUtils.readFromSource(source, null);
+
+    assertThat(expected, containsInAnyOrder(actual.toArray()));
+  }
+
+  @Test
+  public void testSplits() throws Exception {
+    PipelineOptions options = PipelineOptionsFactory.create();
+
+    List<KV<IntWritable, Text>> expected = createRandomRecords();
+    String filename = generateTestFile("tmp.avro", expected);
+
+    HadoopFileSource<IntWritable, Text> source =
+        HadoopFileSource.from(filename, SequenceFileInputFormat.class,
+            IntWritable.class, Text.class);
+
+    // Assert that the source produces the expected records
+    assertEquals(expected, SourceTestUtils.readFromSource(source, options));
+
+    // Split with a small bundle size (has to be at least size of sync interval)
+    List<? extends BoundedSource<KV<IntWritable, Text>>> splits = source
+        .splitIntoBundles(SequenceFile.SYNC_INTERVAL, options);
+    assertTrue(splits.size() > 2);
+    SourceTestUtils.assertSourcesEqualReferenceSource(source, splits, options);
+    int nonEmptySplits = 0;
+    for (BoundedSource<KV<IntWritable, Text>> subSource : splits) {
+      if (SourceTestUtils.readFromSource(subSource, options).size() > 0) {
+        nonEmptySplits += 1;
+      }
+    }
+    assertTrue(nonEmptySplits > 2);
+  }
+
```

Comment:
Some more things to test:
- That readers are closed
- That everything works correctly (including closing readers) if you read a filepattern where some of the matched files are empty
See tests in https://github.com/GoogleCloudPlatform/DataflowJavaSDK/blob/master/sdk/src/test/java/com/google/cloud/dataflow/sdk/io/FileBasedSourceTest.java - the majority are inapplicable, but some may be useful to mimic.